### ISSUE TITLE
[Question] Not sure what is going wrong

### ISSUE THREAD

**Issue by Ghostdreamer11** [Cysharp/UniTask#552]
Hello! I am trying to move a chunk of processing over to a background thread and I am not sure where I am going wrong. Functionally, it's perfect. Untill... things just kind of start to hang? I first noticed when I saw elapsed time going up in the task viewer (even though they are successful). It continues to go up in the task viewer even *after* the player has stopped. 
![image](https://github.com/Cysharp/UniTask/assets/77739872/a296087e-7641-4088-81dd-4c058dd64d3e)

Normally it's just the one ... during processing we may end up showing 7-9 tasks, even though we are explicitly awaiting the completing of the one before starting the other (idk if that still spins up new ones?)
![image](https://github.com/Cysharp/UniTask/assets/77739872/6690a852-f5d5-4a7c-a2ab-dbb66107dc5e)

I am not extremely familiar with async/await but have done a lot of reading and cancellation tokens are something... important. I thought I had it covered, maybe not? Any help would be great!

The code itself does exactly as it should (everything updates, it chews the code, its great). I am just concerned with these hanging threads. This is the only place we are using UniTask, it is only part of a single script, and it is not a prefab/singleton/static/etc.

```
        private UniTask updateMapsTask;
        private void Update()
        {
            if (Time.time - _lastUpdated > updateFrequency)
            {
                _lastUpdated = Time.time;

                Debug.Log($"Status: {updateMapsTask.Status}");
                if(!updateMapsTask.Status.IsCompleted()) return;
                
                updateMapsTask = WorldGrid.UpdateMapsAsync();
            }
        }

        //WorldGrid
        public async UniTask UpdateMapsAsync()
        {
            //await UniTask.RunOnThreadPool(BakeMaps, Application.exitCancellationToken);
            await UniTask.RunOnThreadPool(BakeMaps, true, Application.exitCancellationToken);
        }


        private void BakeMaps()
       {
            //Big ol method that does lots, but doesn't touch the unity API and runs just fine on the background thread
       }
```

We are doing it in update now, but we will end up using timers instead!

**Comment by neuecc** [comment]
UniTask is optimized (used Tasks are returned to the pool) when await or Forget() is called.
Also, avoid direct references to Status as much as possible, and if you need the completion flag, define a separate bool.
Don't try to get it from UniTask.
From the looks of it, UpdateMapsAsync is FireAndForget-like, so WorldGrid.UpdateMapsAsync().Forget(); with a return value of UniTaskVoid seems like a good idea.
Would it be better to have a flag with a bool to determine if it is running or not?

```csharp
State mapUpdateState;

private void Update()
{
    if (!mapUpdateState.Updating && Time.time - mapUpdateState.LastUpdated > updateFrequency)
    {
        WorldGrid.UpdateMapsAsync(mapUpdateState).Forget();
    }
}

public async UniTaskVoid UpdateMapsAsync(State state)
{
    state.Updating = true;
    try
    {
        await UniTask.RunOnThreadPool(BakeMaps, true, Application.exitCancellationToken);
    }
    finally
    {
        state.Updating = false;
        state.LastUpdated = Time.time;
    }
}

class State
{
    public bool Updating;
    public float LastUpdated; 
}
```

**Comment by Ghostdreamer11** [comment]
OK, alright, got it working. 
I can't see the task in the UniTask anymore (because its been properly put on the thread? maybe?) --- but I can safely assume (?) that the thread isnt hanging after leaving, and the previous threads arent holding onto the task (it would be nice to have a way to prove it though). 

Some clarifications (since we will be doing something similar to this in a few places!) .... 

The reason why we were seeing the elapsed time continue up was because on the last iteration of the loop, we called await and spun up the task, but it got interrupted so await was never called again (and therefore it was never returned to the pool)?

If we had this as a non-background thread (coroutine-ish) were we doing it about the right way with the cancellation token?

Conversely, we had messed with Awaitable and got the "same" functionanily using MoveToThreadPool/MoveToMainThread , does UniTask work the same? (We dropped Awaitable because of some other issues, and we didn't want to bump out of LTS)

We did use the flag, if asking for the status directly is no good then we are fine with the flag... we can just reuse the class as we go alone and layer on top of it. 

This package has been great, we are getting ready to dive deep into asyncing a ton of AI processing that isn't jobs friendly but pushing this piece of code (that takes ~3 seconds to complete and critical) has already paid itself over 10x in dev costs. Is there a "donate to support the amazing author of this package" button somewhere? :)

**Comment by neuecc** [comment]
To clarify, the UniTasks that were accumulating in the Tracker were not actually leaking. It's just that they were not being released from the Tracker (awaiting them releases them from the Tracker).

The reason it's not recommended to directly check the Status is because when UniTask is used correctly, it may be returned to the pool. In that case, the status cannot be guaranteed (more precisely, if it has been returned to the pool, the Status check will throw an exception to prevent returning an invalid state).

By the way, the sponsor button is at the top of the page :)
